File: SearchSortedRotated.java

// Time Complexity : O(log(n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :

class Solution {
    public int search(int[] nums, int target) {
        if(nums == null){
            return -1;
        }
        int low = 0;
        int high = nums.length -1;

        while(low<=high){
            int mid = low + (high - low)/2;
            if(nums[mid]== target) return mid;
            if(nums[mid]>=nums[low]){
                if(target >= nums[low] && target < nums[mid]){
                    high = mid-1;
                }else{
                    low = mid +1;
                }
            }
            if(nums[mid]<=nums[high]){
                if(target > nums[mid] && target <= nums[high]){
                    low = mid +1;
                }else{
                    high = mid-1;
                }
            }

        }
        return -1;
        
    }
}

File: searchSortedRotated.cpp
// Time Complexity : O(log(n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
/* One part is always sorted.
Find sorted part and search for element in sorted part.
If not in sorted part, it will be in unsorted part.
*/
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        if(n ==0){
            return -1;
        }
        int l=0;
        int h = n-1;
        int m=0;
        while(l<=h){
            m = l+(h-l)/2;
            if(nums[m]== target){
                return m;
            }
            //left side sorted and element lies in sorted side
            if(nums[l]<=nums[m]){
                if(nums[l]<=target && nums[m]> target){
                    h = m-1;
                }
                else{
                    l = m+1;
                }
            }
            else{
                if(nums[h]>=target && nums[m]< target){
                    l = m+1;
                }
                else{
                    h = m-1;
                }
            }
            

        }
        return -1;
    }
};